hack/k-license/k-license.go

```diff
@@ -0,0 +1,151 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var (
+	templatesDir = flag.String("templates", "../boilerplate", "directory containing license templates")
+	excludeDirs  = flag.String("exclude", ".git,vendor,node_modules", "comma-separated list of directories to exclude")
+	path         = flag.String("path", ".", "Absolute Path to directory location")
+	helpFlag     = flag.Bool("help", false, "Display this help message and exit.")
+)
+
+func main() {
+	flag.Parse()
+	if *helpFlag {
+		printHelp(os.Args[0])
+		return
+	}
+	excludeList := strings.Split(*excludeDirs, ",")
+	err := filepath.Walk(*path, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() && contains(excludeList, info.Name()) {
+			return filepath.SkipDir
+		}
+		if !info.IsDir() && (isCodeFile(path) || isBuildFile(path)) && !hasLicense(path) {
+			currentYear := strconv.Itoa(time.Now().Year())
+			err := addLicense(path, *templatesDir, currentYear)
+			if err != nil {
+				return err
+			}
+			fmt.Printf("Modified %s file\n", path)
+		}
+
+		return nil
+	})
+	if err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		os.Exit(1)
+	}
+}
+
+func contains(list []string, str string) bool {
+	for _, item := range list {
+		if item == str {
+			return true
+		}
+	}
+	return false
+}
+
+func printHelp(progName string) {
+	fmt.Fprintf(os.Stderr, "Usage: %s [--help] [-d dir] [--exclude ] [--templates]\n", progName)
+	fmt.Fprintln(os.Stderr, "Add a copyright/license header to all source files in a Git repository.\n-")
+	fmt.Fprintln(os.Stderr, "Options:")
+	flag.PrintDefaults()
+	fmt.Fprintln(os.Stderr, "\nExamples:")
+	fmt.Fprintf(os.Stderr, "  %s               # Add license headers to all source files in the current directory and its subdirectories, using the default file extensions and templates directory.\n", progName)
+	fmt.Fprintf(os.Stderr, "  %s --dir /path/to/repo --exclude vendor,node_modules --templates /path/to/templates   # Add license headers to all .go, .sh, .py, makefile, docker files in /path/to/repo and its subdirectories, using the license header templates in /path/to/templates.\n", progName)
+}
+
+func isCodeFile(path string) bool {
+	ext := strings.ToLower(filepath.Ext(path))
+	switch ext {
+	case ".go", ".c", ".h", ".cpp", ".py", ".ipynb", ".java":
+		return true
+	}
+	return false
+}
+func isBuildFile(path string) bool {
+	switch {
+	case strings.HasSuffix(path, ".sh"):
+		return true
+	case strings.HasSuffix(path, "Makefile"):
+		return true
+	case strings.HasSuffix(path, "Dockerfile"):
+		return true
+	}
+	return false
+}
```

Comment:
**suggestion:** instead of using switch statements here to check the file extenstion, how about storing the code/build file extensions as maps and then perform check?

That would improve the search and make it easier to maintain/add new file extensions in future?


```
var codeFileExtensions = map[string]bool{
	".go":    true,
	".c":     true,
	".h":     true,
	...
}

var buildFileExtensions = map[string]bool{
	".sh":       true,
	"Makefile":  true,
	"Dockerfile": true,
}

func isCodeFile(path string) bool {
	return codeFileExts[strings.ToLower(filepath.Ext(path))]
}
...

```